### DISCUSSION TITLE
Proposal: Mark a package as following Semantic Versioning

### DISCUSSION THREAD

**Discussion by ghost** [composer/composer#10986] [General]
Semantic versioning has been around for a log time now. We all agree that this is a great way to deliver safe, reliable software and fortunately many package maintainers follow this process.
However, currently this is mostly implicit and a package consumer is not guaranteed that the Semantic Versioning is respected. The best we can get is when maintainer explicitly mentions this in README file or even better, inside CHANGELOG file as the last Guiding Principle of [keep a changelog](https://keepachangelog.com) recommends. Unfortunately I see this very rarely. 
Otherwise, to be 100% sure, all we can do is to go through the changelog file (if one exists, what is not always the case) and read line by line that there aren't any breaking changes between versions. What's more, sometimes the change description might look like a breaking change for the reader who is not familiar with the internal implementation of the package and the confusion begins.
Of course, one could say that just tagging versions in a `x.y.z` manner is enough to assume that package follows Semantic Versioning.  I'm not so sure about that. However I'm not the maintainer of any package so if this is a true assumption please let me know in the comments.

I have the following idea for Composer as the one and only package manager in the PHP World.
**Extend composer.json schema by a new key that says this package follows Semantic Versioning**.
Let's say this key is named `follows_semver`. Then in package's composer.json: 
```json
{
  "follows_semver": true
}
``` 
By default this config key would be `false` (if a maintainer doesn't explicitly set this to `true` we assume he doesn't follow semver).
Then when running `composer update [package]` Composer could emit a warning that this package does not follow Semantic Versioning and the upgrade might break the application (there also could be a question "Do you want to continue?", but that's just an implementation detail).

What do you think?

**Comment by Seldaek** [comment]
IMO this isn't really necessary as the common assumption at this point is that people do follow semver unless otherwise stated. There is a lot of community pressure for this to happen too as it's a pain when someone doesn't, and Laravel for example was kinda forced into it by its community if I recall correctly.

In my personal experience  the few packages not following it are fairly obviously using some different versioning scheme, but I'm sure you can find a few examples that do not.
